Agents/BMSUpdateAgent/BMSUpdateAgent/src/main/java/uk/ac/cam/cares/jps/agent/bmsupdate/BMSUpdateAgent.java

```diff
@@ -144,4 +164,225 @@ private String parseFanStatusMessage(String response) {
+        SelectQuery selectQuery = Queries.SELECT();
+        Variable objectIdVar = SparqlBuilder.var("objectIdVar");
+        TriplePattern getBacnetObjectId = GraphPatterns.tp(iri(dataIRI), P_ONTOBMS.iri("hasBacnetObjectID"), objectIdVar);
+        selectQuery.prefix(P_ONTOBMS).select(objectIdVar).where(getBacnetObjectId);
+        try {
+            LOGGER.info("getting Bacnet Object ID of " + dataIRI +" ...");
+            LOGGER.info("execute query: " + selectQuery.getQueryString());
+            JSONArray results = new JSONArray(rsClient.execute(selectQuery.getQueryString()));
+            return results.getJSONObject(0).getString("objectIdVar");
+        } catch (Exception e) {
+            LOGGER.error(FAIL_TO_GET_BACNETOBJECTID);
+            throw new JPSRuntimeException(FAIL_TO_GET_BACNETOBJECTID);
+        }
+    }
+
+    /**
+     * Get latest timeseries value for the data IRI and compared it with the trigger value,
+     * if they are equal, run a sparql update to insert a set of triples and delete a set of triples
+     * @param dataIRI The data iri
+     * @param rsClient The knowledge graph client
+     * @param RDBClient The remote RDB client
+     * @param triggerValue The value to compare with the latest timeseries value
+     * @param insertString The string to insert via sparql update
+     * @param deleteString The string to delete via sparql update
+     * @return 
+     */
+    public JSONObject checkInsertAndDelete(String dataIRI, RemoteStoreClient rsClient, RemoteRDBStoreClient RDBClient, String triggerValue, String insertString, String deleteString) {
+        JSONObject result = new JSONObject();
+        tsClient = new TimeSeriesClient<>(rsClient ,OffsetDateTime.class);
+        try (Connection conn = RDBClient.getConnection()) {
+            timeseries = tsClient.getLatestData(dataIRI, conn);
+        } catch (Exception e) {
+            throw new JPSRuntimeException(FAIL_TO_GET_LATESTDATA + dataIRI);
+        }
+        String latestValue = timeseries.getValuesAsString(dataIRI).get(timeseries.getValuesAsString(dataIRI).size() - 1);
+        OffsetDateTime latestTimeStamp = timeseries.getTimes().get(timeseries.getTimes().size() - 1);
+        Date date = new java.util.Date(latestTimeStamp.toEpochSecond()*1000);
+        SimpleDateFormat sdf = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss a z");
+        sdf.setTimeZone(TimeZone.getDefault());
+        Object ts = sdf.format(date);
+        LOGGER.info("The latest value for " + dataIRI +" is " + latestValue + " and the corresponding timestamp is " + ts.toString());
+        if (result.has("message")){
+            result.accumulate("message", "The latest value for " + dataIRI +" is " + latestValue + " and the corresponding timestamp is " + ts.toString());
+        } else {
+            result.put("message", "The latest value for " + dataIRI +" is " + latestValue + " and the corresponding timestamp is " + ts.toString());
+        }
```

Comment:
Is it possible for result to have "message" key already? It is created at line 226 and first used as line 240.

Reply:
I added this for debugging as well and forgot to remove it after the result format was finalized. I have removed these lines as well.